### ISSUE TITLE
Nightly build docker-compose script not working on Windows platform

### ISSUE THREAD

**Issue by tingyuz** [edgexfoundry/developer-scripts#188]
The latest nightly build docker-compose adds /tmp volume, which is not working in the environment. As a result, security-secrets-setup will be failed to create CA as well KONG/Vault certs under /tmp/secrets/edgexvault. So the script is not working in the Windows.

Possible root cause is from this line: 
- /tmp/edgex/secrets/ca:/tmp/edgex/secrets/ca:ro

**Comment by jim-wang-intel** [comment]
More information: 
It is regarding on how native Windows's path interprets the drive letter.  On native windows it is C:\\tmp\\edgex\\secrets\\ca to be the host name path.  And if using cgywin, it will need to be /cgydrive/c/tmp/edgex/secrets/ca to be the host name path.

**Comment by tingyuz** [comment]
I did another round test on all platforms and notice that the cgywin path is not working if c:/tmp/edge/secrets folder doesn't exist. So to summarize the result:

1. Running docker-compose within windows command window on Windows platform:
Don't need to create c:\\tmp\\edgex\\secrets, and use the format of this will work: c:\\tmp\\edgex\\secrets:/tmp/edgex/secrets      

2. Runing docker-compose file within cygwin on Windows platform:
Need to create c:\\tmp\\edgex\\secrets folder first, then using this format:
/cygdrive/c/tmp/edgex/secrets:/tmp/edgex/secrets      

3. Running docker-ompose file on Linux:
don't need to create /tmp/edgex/secrets, and using this format:
/tmp/edgex/secrets:/tmp/edgex/secrets

**Comment by jim-wang-intel** [comment]
@tingyuz the assumption of deploydir is it has to be existing and writable first before running the application.  See the details here: https://github.com/edgexfoundry/edgex-go/blob/master/internal/security/secrets/helper/helper.go#L303-L317

**Comment by tingyuz** [comment]
@jim-wang-intel do we have script to create such folder in advance before using it? I seem not able to find it.

**Comment by jim-wang-intel** [comment]
> 
> 
> @jim-wang-intel do we have script to create such folder in advance before using it? I seem not able to find it.

No.  The end user creates it first- that is the assumption at that time when design.  We can change this assumption but that is a new change of requirements

**Comment by mkbhanda** [comment]
Thank you @tingyuz and @jim-wang-intel for the analysis. Given we want docker-compose to magically bring up all the services, and towards that the dependency tracking scripts .. it makes sense to ask part of the docker-compose to create the directory. To @tingyuz point of patch varying depending on launch context, can one have variables in docker-compose like in bash scripts. Should the variants be specified in a config file?

**Comment by jim-wang-intel** [comment]
> 
> Thank you @tingyuz and @jim-wang-intel for the analysis. Given we want docker-compose to magically bring up all the services, and towards that the dependency tracking scripts .. it makes sense to ask part of the docker-compose to create the directory. To @tingyuz point of patch varying depending on launch context, can one have variables in docker-compose like in bash scripts. Should the variants be specified in a config file?

Ok, thanks for the description, @mkbhanda.  So, 2 things:

1. What you described above is an asking for changes in the current behavior of `security-secrets-setup` on `deployDir` creation if not exists.  Currently it is not doing that like I mentioned in the previous conversation.  As far as the configuration is concerned , the `deployDir` is already part of configuration toml file.  See this: https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secrets-setup/res/configuration.toml#L8-L10.  

2. For the other part, the variation path on how docker runs on Windows platform, I think one solution is that maybe having a shell script to export the environment variable of that path on the host system and uses that environment variable in the docker-compose file.   However, the end user might need to do some interaction here, as I do not know whether the script can detect the docker-compose up is running on native Winodws or CYGWIN kind of environment and so this may still rely on the end user to tweak that shell script a bit.

cc: @bnevis-i , @tingyuz, @eno-intel

**Comment by bnevis-i** [comment]
`/tmp/edgex/secrets` is a mount-point INSIDE OF THE CONTAINER.  There is nothing that `security-secrets-setup` can to do fix the cross-platform issues.  If Docker doesn't create that path inside of the container, then things will have failed even before `security-secrets-setup` has started running.

The purpose of using /tmp is because it is a white-listed path on MacOS docker, but on a Linux system we really want to use /run in order to avoid writing secrets to the disk.  The other alternative that I see is create a secrets-service1 volume, a secrets-service2 volume, a secrets-service3 volume, etc, but this goes very far away from just doing a search-and-replace of `/tmp/edgex` with `/run/edgex`.

A better alternative to Cygwin is Docker for Windows with Windows Subsystem for Linux version 2.   I have not tested this.

**Comment by bnevis-i** [comment]
@tingyuz Very nice writeup on how the host-portion of the volume mount is interpreted on the various platforms.

**Comment by bnevis-i** [comment]
https://stackoverflow.com/questions/3104410/identify-cygwin-linux-windows-using-environment-variables suggests one could use `$OSTYPE`  to detect which type in a wrapper script and set an appropriate variable.  I don't think it is possible to fix it in the `docker-compose` alone.

**Comment by jim-wang-intel** [comment]
Update: I tested this with @rsdmike on Windows 10 Power Shell.  We found out that it appears not an issue. The docker-compose with `/tmp/edgex/secrets/:/tmp/edgex/secrets` actually working on docker on Windows.   

The reason we thought that it was an issue is that we could not "see" the physical directory or folder created but actually docker on Windows somehow does.   It is something related to how docker works on Windows regarding to /dir/path/name.  It is not the same to assume that the root / is equivalent to c:/  . So in short, on docker of windows, c:/ is not the same as / like we thought it was. Also if you docker exec into the particular container of `security-secrets-setup`, you can also find the generated folders and files.

To prove this, we actually change `/tmp/edgex/secrets` to `./tmp/edgex/secrets` and it actually does create three sub folders (`ca`, `edgex-vault`, `edgex-kong`) and one can now actually "see" the creation of cert files.

So in short, I think we can disposition this issue.

**Comment by bnevis-i** [comment]
I don't know if the code still works, but there is a blog post that shows how to get into the shadow VM created by Docker for Windows.  Maybe we could use this to validate for sure that there isn't an issue prior to closing it out.

https://blog.jongallant.com/2017/11/ssh-into-docker-vm-windows/

**Comment by bnevis-i** [comment]
Assigned back to TingYu for validation.

**Comment by bnevis-i** [comment]
Awaiting validation by @tingyuz .

**Comment by tingyuz** [comment]
Verified on Power Shell environment and confirmed the latest docker-compose file works on Windows.  
After verifying on power shell, I reset the environment and run the docker-compose again within cygwin. To my surprise it works well. I checked c:\tmp folder as well as C:\cygwin64\tmp\edgex\secrets but don't see any cert files there. So it could be in some hidden location as Bryon and Jim mentioned. Anyway we have an option of running the script within Power Shell, I believe we can just add this into release note or user guide if necessary and it is OK to close the issue.
